In [16]:
import tensorflow as tf
import numpy as np

In [1]:
### Loading the dataset and storing it
from tensorflow.keras.datasets import mnist
(X_train, y_train), (X_test, y_test) = mnist.load_data()
### Shape of the datast
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(60000, 28, 28) (10000, 28, 28) (60000,) (10000,)


In [2]:
### Type of the data
print(X_train.dtype,y_train.dtype)

uint8 uint8


In [3]:
### Reshape the train and test data according to the model requirements
X_train = X_train.reshape([-1, 28, 28, 1]).astype('float32')
X_test = X_test.reshape([-1, 28, 28, 1]).astype('float32')

X_train_scaled = X_train/255
X_test_scaled = X_test/255

In [4]:
### One hot encoding the target as it is a classification problem

from tensorflow.keras.utils import to_categorical
Y_train_req = to_categorical(y_train)
Y_test_req = to_categorical(y_test)
print(Y_train_req.shape, Y_test_req.shape)

(60000, 10) (10000, 10)


In [5]:
from tensorflow.keras.layers import Conv2D, Dense, Flatten, BatchNormalization, Dropout, Input, MaxPooling2D, MaxPool2D
from tensorflow.keras.models import Sequential

### LeNet-5 Model from scratch (slighly modified some layers)

In [8]:
LeNet5 = Sequential()
LeNet5.add(Conv2D(6, (5,5), strides=1, input_shape=(28,28,1), activation='relu'))
LeNet5.add(MaxPool2D((2,2), strides=2))
LeNet5.add(Conv2D(16, (5,5), strides=1, activation='relu'))
LeNet5.add(MaxPool2D((2,2), strides=2))
LeNet5.add(Flatten())
LeNet5.add(Dense(120, activation='relu'))
LeNet5.add(Dense(84, activation='relu'))
LeNet5.add(Dense(10, activation='softmax'))

In [9]:
LeNet5.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_2 (Conv2D)           (None, 24, 24, 6)         156       
                                                                 
 max_pooling2d_2 (MaxPoolin  (None, 12, 12, 6)         0         
 g2D)                                                            
                                                                 
 conv2d_3 (Conv2D)           (None, 8, 8, 16)          2416      
                                                                 
 max_pooling2d_3 (MaxPoolin  (None, 4, 4, 16)          0         
 g2D)                                                            
                                                                 
 flatten_1 (Flatten)         (None, 256)               0         
                                                                 
 dense_3 (Dense)             (None, 120)              

In [10]:
LeNet5.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [11]:
LeNet5.fit(X_train, Y_train_req, validation_split=0.1, epochs=15, batch_size=32)

Epoch 1/15
1688/1688 [==============================] - 27s 15ms/step - loss: 0.4617 - accuracy: 0.9151 - val_loss: 0.1165 - val_accuracy: 0.9673
Epoch 2/15
1688/1688 [==============================] - 27s 16ms/step - loss: 0.0954 - accuracy: 0.9715 - val_loss: 0.0784 - val_accuracy: 0.9770
Epoch 3/15
1688/1688 [==============================] - 26s 15ms/step - loss: 0.0724 - accuracy: 0.9779 - val_loss: 0.0684 - val_accuracy: 0.9793
Epoch 4/15
1688/1688 [==============================] - 27s 16ms/step - loss: 0.0616 - accuracy: 0.9817 - val_loss: 0.0547 - val_accuracy: 0.9848
Epoch 5/15
1688/1688 [==============================] - 28s 17ms/step - loss: 0.0511 - accuracy: 0.9843 - val_loss: 0.0519 - val_accuracy: 0.9855
Epoch 6/15
1688/1688 [==============================] - 28s 16ms/step - loss: 0.0490 - accuracy: 0.9855 - val_loss: 0.0497 - val_accuracy: 0.9842
Epoch 7/15
1688/1688 [==============================] - 26s 16ms/step - loss: 0.0406 - accuracy: 0.9873 - val_loss: 0.0622 -

In [13]:
LeNet5.evaluate(X_test, Y_test_req, batch_size=32)

313/313 [==============================] - 3s 9ms/step - loss: 0.0557 - accuracy: 0.9865


[0.055666692554950714, 0.9865000247955322]

In [19]:
from sklearn.metrics import confusion_matrix

# get probabilities
probabilities = LeNet5.predict(X_test)

# index of the class with the highest probability for each sample
predictions = np.argmax(probabilities, axis=1)

### Printing the confusion Matrix
print('CONFUSION MATRIX')
print(confusion_matrix(y_test, predictions))


313/313 [==============================] - 2s 5ms/step
CONFUSION MATRIX
[[ 954    0    2    0    0    1   18    1    3    1]
 [   0 1130    3    0    0    0    0    0    1    1]
 [   0    0 1027    1    0    0    0    3    1    0]
 [   0    0    0 1005    0    2    0    3    0    0]
 [   0    0    0    0  960    0    6    1    0   15]
 [   0    0    0    8    0  875    5    1    1    2]
 [   1    3    0    0    0    2  948    0    4    0]
 [   0    2   12    1    0    0    0 1009    0    4]
 [   0    0    1    3    0    0    1    1  966    2]
 [   0    0    1    1    4    4    1    2    5  991]]


### AlexNet model architecture from scratch

In [30]:
### Creating The AlexNet Model
AlexNet = Sequential()
AlexNet.add(Conv2D(96, (3,3), strides=1, input_shape=(28,28,1), activation='relu'))
AlexNet.add(MaxPool2D((3,3), strides=2))
AlexNet.add(Conv2D(256,(5,5), padding='same', activation='relu'))
AlexNet.add(MaxPool2D((3,3), strides=2))
AlexNet.add(Conv2D(384,(3,3), padding='same', activation='relu'))
AlexNet.add(Conv2D(384,(3,3), padding='same', activation='relu'))
AlexNet.add(Conv2D(256,(3,3), padding='same', activation='relu'))
AlexNet.add(MaxPool2D((3,3), strides=2))
AlexNet.add(Flatten())
AlexNet.add(Dense(4096, activation='relu'))
AlexNet.add(Dense(4096, activation='relu'))
AlexNet.add(Dense(10, activation='softmax'))

In [31]:
AlexNet.summary()

Model: "sequential_5"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_19 (Conv2D)          (None, 26, 26, 96)        960       
                                                                 
 max_pooling2d_13 (MaxPooli  (None, 12, 12, 96)        0         
 ng2D)                                                           
                                                                 
 conv2d_20 (Conv2D)          (None, 12, 12, 256)       614656    
                                                                 
 max_pooling2d_14 (MaxPooli  (None, 5, 5, 256)         0         
 ng2D)                                                           
                                                                 
 conv2d_21 (Conv2D)          (None, 5, 5, 384)         885120    
                                                                 
 conv2d_22 (Conv2D)          (None, 5, 5, 384)        

In [32]:
AlexNet.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [33]:
AlexNet.fit(X_train, Y_train_req, validation_split=0.1, epochs=5, batch_size=32)

Epoch 1/5
1688/1688 [==============================] - 2182s 1s/step - loss: 0.7149 - accuracy: 0.9239 - val_loss: 0.1249 - val_accuracy: 0.9668
Epoch 2/5
1688/1688 [==============================] - 2179s 1s/step - loss: 0.1106 - accuracy: 0.9713 - val_loss: 0.0980 - val_accuracy: 0.9747
Epoch 3/5
1688/1688 [==============================] - 2175s 1s/step - loss: 0.0924 - accuracy: 0.9770 - val_loss: 0.1448 - val_accuracy: 0.9685
Epoch 4/5
1688/1688 [==============================] - 2162s 1s/step - loss: 0.0787 - accuracy: 0.9796 - val_loss: 0.0689 - val_accuracy: 0.9827
Epoch 5/5
1688/1688 [==============================] - 2212s 1s/step - loss: 0.0700 - accuracy: 0.9832 - val_loss: 0.0536 - val_accuracy: 0.9868


In [34]:
AlexNet.evaluate(X_test, Y_test_req, batch_size=32)

313/313 [==============================] - 73s 234ms/step - loss: 0.0633 - accuracy: 0.9853


[0.06333968788385391, 0.9853000044822693]

In [35]:
# get probabilities
probabilities = LeNet5.predict(X_test)

# index of the class with the highest probability for each sample
predictions = np.argmax(probabilities, axis=1)

### Printing the confusion Matrix
print('CONFUSION MATRIX')
print(confusion_matrix(y_test, predictions))

313/313 [==============================] - 4s 12ms/step
CONFUSION MATRIX
[[ 954    0    2    0    0    1   18    1    3    1]
 [   0 1130    3    0    0    0    0    0    1    1]
 [   0    0 1027    1    0    0    0    3    1    0]
 [   0    0    0 1005    0    2    0    3    0    0]
 [   0    0    0    0  960    0    6    1    0   15]
 [   0    0    0    8    0  875    5    1    1    2]
 [   1    3    0    0    0    2  948    0    4    0]
 [   0    2   12    1    0    0    0 1009    0    4]
 [   0    0    1    3    0    0    1    1  966    2]
 [   0    0    1    1    4    4    1    2    5  991]]


#### The validation accuracy using LeNet5 architecture is 0.9865 and using AlexNet architecture is 0.9853